Column resizing

dnorth opened the pull request:
Closes #2555

## ✅ Pull Request Checklist:

- [ ] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [ ] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## 📝 Test Instructions:

- storybook tests with 'allowsResizing' prop in TableView
- npm run test Table.test.js

## TODOs as followup potentially
- Move resizers up to be siblings of the rows such that their hit area can cover the columns and the focus ring is visible all the way around
- Get rid of the timeout to move focus from the menutrigger to the resizer, right now it's dependent on the menutrigger close timeout of ~350ms when the focus is restored to the trigger element


## 🧢 Your Project:

Adobe Analytics (Analysis Workspace)

dnorth commented on packages/@react-spectrum/table/src/TableView.tsx:
```diff
@@ -18,11 +18,29 @@ import {FocusRing, useFocusRing} from '@react-aria/focus';
 import {GridNode} from '@react-types/grid';
```
This is the one file we need to figure out how to undo our prettier changes, sorry! We'll try to figure that out soon. Unfortunately it feels like we'll have to do this manually unless someone has a better idea.

snowystinger replied:
without knowing how many actual changes you made, I don't really know, you could copy our existing code and just apply your changes in manually if there aren't too many
I assume it's trying to fit the 80char column scheme?

dnorth replied:
That's probably a good idea, @snowystinger and yes you're probably right about the 80char column scheme.

dnorth commented on packages/@react-spectrum/table/test/Table.test.js:
```diff
@@ -4259,9 +4259,9 @@ describe('TableView', function () {
-          expect(row.childNodes[3].style.width).toBe('320.6666666666667px');
+          expect(row.childNodes[1].style.width).toBe('320.67px');
+          expect(row.childNodes[2].style.width).toBe('320.66px');
+          expect(row.childNodes[3].style.width).toBe('320.67px');
```
Currently rounding to 2 decimal places for predictability. We're seeing a visual bug that might be fixed if we just round to whole numbers though, so we are considering doing that if there's no reason to be rounding to 2 decimal places and if it actually does fix the bug.

dnorth replied:
Okay we may just round then. I'll test it out when I have a chance. We were rounding initially and then realized that tests failed because they were expecting explicit non-rounded pixel results so we wondered if there was a reason that rounding wasn't happening. I think as far as pixels go though, rounding makes the most sense.

dnorth commented on packages/@react-stately/layout/src/TableLayout.ts:
```diff
@@ -63,48 +65,87 @@ export class TableLayout<T> extends ListLayout<T> {
     ];
   }
 
-  buildColumnWidths() {
```
The big change here is that we removed most all of the logic specifically for building column widths and put it into the `useColumnResizeWidthState` hook

dnorth commented on packages/@react-stately/slider/src/useSliderState.ts:
```diff
@@ -165,12 +165,16 @@ export function useSliderState(props: SliderStateOptions): SliderState {
 
   const [values, setValues] = useControlledState<number[]>(
```
prettier changes. Ignore for now, we'll fix these.

dnorth commented on packages/@react-stately/table/src/useColumnResizeWidthState.ts:
```diff
@@ -0,0 +1,227 @@
+
```
Some of this is functionality pulled out from TableLayout and some is functionality that we added. 

If you would like to see a simplified version of the algorithm we wanted to use here then you can view this:
https://codesandbox.io/s/column-width-resizer-thingy-danny-fork-jo4rr?file=/src/column-utils.js

You can note the algorithm with comments and the spec file

dnorth commented on packages/@react-stately/virtualizer/src/Virtualizer.ts:
```diff
@@ -321,6 +321,9 @@ export class Virtualizer<T extends object, V, W> {
   private _renderView(reusableView: ReusableView<T, V>) {
+    if (!reusableView.layoutInfo) {
+      return;
+    }
```
We were running into issues where layoutInfo wasn't available and it would cause errors. It makes sense that if layoutInfo isn't available then we should just return, but this is one possible section where we're avoiding some kind of deeper underlaying problem and if you have any ideas on why this might be happening and how we should fix it instead then we'd be happy to work together through this.

dnorth replied:
I believe the "loads more on scroll when contentSize.height < rect.height * 2" storybook test is one that fails if you remove these changes - if someone wanted to test the behavior.

snowystinger reviewed (commented):
Hey! I know you're still cleaning some things up so hopefully this initial review doesn't conflict with what you're doing. Great work so far.

I think the only usage note I have right now is that while expanding a column, the horizontal scroll bar sometimes appears, and this makes tons of sense.

When shrinking columns though, it's possible to make the table take up less space than is allocated for it which leaves a blank area on the far right in LTR.

Some questions about resize behavior:
Should the last column be resizable?
If the combined widths of columns is less than the width of the table, should the last column snap to the edge of the table area?
Should you be able to resize a column by dragging off the table? I noticed that if I drag back onto the table it'll stop the resizing even if I haven't let off the mouse.

Some general questions:
How big of a hit area should the resizer have? it can be a bit tricky to grab.
Can we have some stories with the prop `showDivider`? I think the whole column divider is supposed to become blue? should it all be grabbable for resizing?

Again, great job so far. It's honestly working pretty well already.

dnorth commented on packages/@react-stately/table/src/useTableState.ts:
```diff
@@ -35,9 +55,16 @@ export interface CollectionBuilderContext<T> {
-  showSelectionCheckboxes?: boolean
+  showSelectionCheckboxes?: boolean,
+  /** Function for determining the default width of columns. */
+  getDefaultWidth: (props) => string | number,
```
We actually realized this would be a breaking API change if we didn't for anyone using useTableState and so we did change that to be optional down the chain.

dnorth commented on packages/@react-aria/table/src/useTableColumnResize.ts:
```diff
@@ -0,0 +1,40 @@
+  const {moveProps} = useMove({
+    onMoveStart() {
+      // stateRef.current.addResizedColumn(item.key);
+      columnResizeWidthRef.current = stateRef.current.getColumnWidth(item.key);
```
This was one question we had. We only added onEnd because of slider but I think it would make sense to add an onStart.

dnorth replied:
Although the only information you can gleam from an onStart is the fact that it's starting, so maybe it's not as helpful. The onEnd will return the full information. The column resizing information doesn't make sense onStart

dnorth commented:
In response to your comment above, @snowystinger which for some reason I can't reply to directly...

> I think the only usage note I have right now is that while expanding a column, the horizontal scroll bar sometimes appears, and this makes tons of sense.

- Yes, and we actually just pushed out a fixed to a bug with this where if you resize when there's a scrollbar in the overflow area, it was acting strange. I'm not sure if you ever tried that, but if you did it should be fixed now.

> When shrinking columns though, it's possible to make the table take up less space than is allocated for it which leaves a blank area on the far right in LTR.

- Yes that is very true! This is intentional but I can see why someone might think it's strange. We actually explored examples of how this behavior would work and a popular paid-for table view, ag-grid, does this same behavior (https://ag-grid.com/javascript-data-grid/column-sizing/)

 > Some questions about resize behavior: Should the last column be resizable? If the combined widths of columns is less than the width of the table, should the last column snap to the edge of the table area? Should you be able to resize a column by dragging off the table?
 
 - These are some good questions and I do think that the way we have it built now feels good and works well. We did talk extensively with Alan Wilson about the different behaviors of the table, including the questions you have above, and he didn't have any concerns with the way we were building it. That being said, something like "don't allow the last column to be resized" isn't a very hard thing to change so if we decide we want to change it then we can have more discussion about it in the future.
 
>  I noticed that if I drag back onto the table it'll stop the resizing even if I haven't let off the mouse.

- This was the bug I referenced above which should be fixed now. Looks like you did see it.

 > Some general questions: How big of a hit area should the resizer have? it can be a bit tricky to grab.
 
 We may have to go back and look at this. I THINK it's supposed to be 4px to the left and 4px to the right, making an 8px range. However because of how the boundaries of CSS work in RSP - I think making the 4px to the right isn't something we did. We may need help with that part of it if we want to make it bigger. 
 
 >  Can we have some stories with the prop `showDivider`? 
 
 - We can definitely add a `showDivider` story, that's a good idea. Or at least add it as a prop to some of the tests. 
 
 > I think the whole column divider is supposed to become blue? should it all be grabbable for resizing?
 
 - This is one thing we will want to work with your team about, since you probably have better knowledge of the codebase. You're right in saying that the design spec has the whole column divider becoming blue, but with how the contains boundaries for CSS work in RSP, we decided to keep it just at the header level for now and work with your team about what it would take to make it go across everything. It's possible you have a good idea here.

dnorth commented:
Before we added to this codebase, @marshallpete and I created multiple artifacts to help define the functionality of the column resizing and prototype it out to understand all the nuances of the problem. Since you're looking over this now @snowystinger , it may be helpful for you to look through our documents to understand how some of these pieces are meant to work. We created these artifacts a few months ago, so it's possible that they are somewhat out of date now, but I think it will still be useful for you and anyone else seeing this.

Column Resizing Recommendation Document:

[Column resizing recommendations.docx](https://github.com/adobe/react-spectrum/files/8137310/Column.resizing.recommendations.docx)

Column Resizing Flow-chart:
[Columns Flow.pdf](https://github.com/adobe/react-spectrum/files/8137323/Columns.Flow.pdf)

(Less useful now that this is in the codebase itself but here are prototypes)

Codesandbox - shows the algorithm we used for handling DYNAMIC column resizing. Assumes there are no STATIC widths. This also includes all the test-cases we thought relevant to the algorithm. These test cases are currently being covered in `Table.test.js`

 https://codesandbox.io/s/column-width-resizer-final-jo4rr?file=/src/column-utils.spec.js

Codepen - This code snippet and demonstration shows the basic behavior of how the columns should resize following the rules defined in the first document. It's probably not as helpful now since you can play with the full behavior in the RSP codebase itself. 

https://codepen.io/mpeterson/pen/VwzRQJo

dnorth commented on packages/@react-spectrum/table/src/Resizer.tsx:
```diff
@@ -0,0 +1,13 @@
+import {useTableColumnResize} from '@react-aria/table/src/useTableColumnResize';
+
+
+export default function Resizer(props) {
```
Okay so for this I was just looking at it. So the example you gave is a hook `useTableCell` but the component here is just a `Resizer` component and so my understanding is that the first argument is always the props and the second argument will only be relevant when using ref forwarding. Does that sound right?

dnorth commented on packages/@react-spectrum/table/src/TableView.tsx:
```diff
@@ -417,6 +431,16 @@ function TableColumnHeader({column}) {
+
+  // Can these be enforced with prop type validation?
+  if (columnProps.width && columnProps.allowsResizing) {
+    throw new Error('Controlled state is not yet supported with column resizing. Please use defaultWidth for uncontrolled column resizing or remove the allowsResizing prop.');
```
So if someone does this it will not work as expected, the entire component may as well be broken. I think my main concern with moving this to a `.warn()` is that in order to utilize `allowsResizing` people will have to change their API from their current `width` prop to `defaultWidth`. Most devs will probably not realize this distinction unless they're forced to. 

If you think a warning in this situation is enough, then I can change it to a warning, but just know that the component will not work as a dev intends if they go down this route.

dnorth commented on packages/@react-spectrum/table/src/TableView.tsx:
```diff
@@ -417,6 +431,16 @@ function TableColumnHeader({column}) {
+  }
+
+  if (columnProps.width && columnProps.defaultWidth) {
+    throw new Error('Column props can only support either width or defaultWidth.');
```
I think that's fair.

dnorth commented on packages/@react-stately/table/src/useColumnResizeWidthState.ts:
```diff
@@ -0,0 +1,227 @@
+      Math.min(roundWidth(newWidth), getMaxWidth(column.props.maxWidth))
+    ));
+
+    // keep track of all columns that have been seized
```
I think double clicking to fit content (like excel) would be awesome. It sounds like a really hard problem though because we have to find out the exact size that the largest of the columns' cells content wants to be. We assumed there's no easy way to do that and be performant but maybe you know of a good way to do it? I think it would be a great nice-to-have.

dnorth commented on packages/@react-stately/layout/src/TableLayout.ts:
```diff
@@ -52,9 +52,10 @@ export class TableLayout<T> extends ListLayout<T> {
-    this.buildColumnWidths();
     let header = this.buildHeader();
     let body = this.buildBody(0);
+    this.stickyColumnIndices = this.collection.columns.filter(c => c.props.isSelectionCell || this.collection.rowHeaderColumnKeys.has(c.key)).map(c => c.index);
```
This used to happen with the width calculation but we think it still belongs here and it doesn't have to happen with the width calculation at all, I think it was previously just using the convenient loop that the width calculation provided.

dnorth commented:
Hey @snowystinger - I made changes that I believe addresses all of your comments/recommendations.

As far as this PR the only thing I think that's left (from my understanding) is to make is so that the `allowsResizing` prop goes down the new code path and that if it's not provided then it goes down the old code path; like we talked about last week.

Are there any other things besides that, that you think needs to be changed?

dnorth commented:
As of now we've completed everything we talked about and made changes based on every question/concern that @snowystinger had.

As such, i'm changing this from a draft to an official PR to be reviewed for merging.

Our latest commit includes the code "feature flag" based on the `allowsResizing` prop.

dnorth commented on packages/@react-spectrum/table/test/Table.test.js:
```diff
@@ -4433,7 +4433,7 @@ describe('TableView', function () {
           let tree = render(
             <TableView aria-label="Table">
               <TableHeader>
```
These two tests we added we actually identified as tests that would fail in the current table width layout code. They will work correctly with our new resizing code. 

Therefore for now we'll just add the "allowsResizing" prop to make sure they go through the correct path, but that prop could be removed in the future if you wanted to when the feature flag prop no longer exists.

dnorth commented on packages/@react-stately/layout/src/TableLayout_DEPRECATED.ts:
```diff
@@ -96,7 +96,7 @@ export class TableLayout_DEPRECATED<T> extends ListLayout<T> {
         let props = column.props as ColumnProps<T>;
         let minWidth = props.minWidth != null ? this.parseWidth(props.minWidth) : 75;
         let maxWidth = props.maxWidth != null ? this.parseWidth(props.maxWidth) : Infinity;
-        let width = Math.max(minWidth, Math.min(maxWidth, columnWidth));
```
This is one legitimate change i'd like to make to the "old" code. Right now tests are relying on partial pixel values to pass. Rather than change tests to pass for both the feature-flagged and non-feature-flagged version, I do think it makes sense to keep the widths as whole pixels and not allow for partial pixels.

I believe Rob said they noticed a visual bug with dividers (most likely cause by partial pixels) and so this should address that and keep everything consistent between the old and new.

snowystinger commented on packages/@react-stately/layout/src/TableLayout.ts:
```diff
@@ -10,29 +10,29 @@
   collection: TableCollection<T>;
   lastCollection: TableCollection<T>;
-  columnWidths: Map<Key, number>;
+  getColumnWidth_: (key: Key) => number;
```
something public probably shouldn't have a trailing underscore, what's the reason for it?

dnorth replied:
This is a good call. There was already a `getColumnWidth` used internally and when we were developing in here we didn't want to ruin the behavior because at the time it was being called externally. Now that we've swapped it out, i'll just swap the internal call to use the underscore and let the external call be without underscore 👍

snowystinger commented on packages/@react-spectrum/table/src/TableView.tsx:
```diff
@@ -690,4 +714,28 @@ function CenteredWrapper({children}) {
+  When ready to remove this feature flag, you can remove this whole section of code, delete the _DEPRECATED files, and just replace the export with the _TableView above.
+*/
+function FeatureFlaggedTableView<T extends object>(props: SpectrumTableProps<T>, ref: DOMRef<HTMLDivElement>) {
+  let state = useTableState({
```
Just an FYI for other reviewers, this is indeed a second copy of Table State, used purely to determine if anything is trying to allow resizing

Question @dnorth, if someone is conditionally putting allowsResizing on a column, will that be a problem? one way this could happen is if there is a column with resizing where the entire column is removed. We have a story that removes columns.

marshallpete replied:
The bug we were seeing is fixed by this commit: https://github.com/adobe/react-spectrum/pull/2883/commits/3bf4357067413227be473f50067e1a557d6abc93

snowystinger reviewed (commented):
partial review

dnorth commented on packages/@react-stately/table/src/useTableColumnResizeState.ts:
```diff
@@ -0,0 +1,137 @@
+import {Key, MutableRefObject, useCallback, useRef, useState} from 'react';
+import {useLayoutEffect} from '@react-aria/utils';
+
+type useTableColumnResizeType<T> = [
+  columnWidths: MutableRefObject<Map<Key, number>>,
+  resizeColumns: (column: GridNode<T>, newWidth: number, doneResizing?: boolean) => { key: Key, width: number }[],
+  setTableWidth: (width: number) => void,
+  onColumnResizeStart: () => void,
+  onColumnResizeEnd: () => void
+]
```
So I think there is a problem with doing it this way. This type is actually for the return value of `useTableColumnResizeState`, which is following the tuple-styled return value of `[something, setSomething] = useState()`. So we wouldn't want to change it to an object because we're not trying to destructure from an object. 

These used to be inline but they were really vague as a result. So to make it more clear what the types are of what we're trying to return, I added this type and named each tuple value - even though the names don't actually do or mean anything. It's just there to help understand the type and nothing more.

Thoughts?

snowystinger replied:
I think that'd make more sense if there was only one argument to useTableColumnResizeState

`let [something, setSomething] = useState(defaultSomething)`

but this hook is taking two pieces of state, and returns 1 output state, 2? setters, and 2 onEvents
so it doesn't much feel like useState even though it says state in its name

it feels much more like how `useTableState` works, so I think I'd prefer it to take a single object for the argument with the named keys `columns` and `getDefaultWidth`
and returns an object with the named properities on it as you've already named them

snowystinger replied:
out of curiosity, why put onColumnResize/End both inside and outside of the useTableColumnResizeState, maybe it's better to do something like this

```
let {
  columnWidths,
  setTableWidth,
  onColumnResizeStart,
  onColumnResize,
  onColumnResizeEnd
} = useTableColumnResizeState({
  columns: collection.columns,
  getDefaultWidth,
  onColumnResizeEnd,
  onColumnResize
});
```

which could be simplified in the calling to

```
let columnResizingState = useTableColumnResizeState({
  columns: collection.columns,
  ...props
});
return {
  collection,
  disabledKeys,
  // ...
  ...columnResizingState
}
```

dnorth replied:
That's fair. I'll look into it.

dnorth replied:
> out of curiosity, why put onColumnResize/End both inside and outside of the useTableColumnResizeState

We do need to return these from the `useTableState` so that way `useTableColumnResize` can utilize it.

snowystinger replied:
Talked in person and resolved this

snowystinger commented on packages/@react-stately/layout/src/TableLayout.ts:
```diff
@@ -183,12 +127,13 @@ export class TableLayout<T> extends ListLayout<T> {
 
-  getColumnWidth(node: GridNode<T>) {
+  // used to get the column widths when rendering to the DOM
+  getColumnWidth_(node: GridNode<T>) {
```
maybe `getAggregateColumnWidth` would make more sense? `getColumnSpanWidth`?

snowystinger commented on packages/@react-stately/layout/src/TableLayout.ts:
```diff
@@ -183,12 +127,13 @@ export class TableLayout<T> extends ListLayout<T> {
     for (let i = 0; i < colspan; i++) {
       let column = this.collection.columns[node.index + i];
-      width += this.columnWidths.get(column.key);
+      width += this.getColumnWidth(column.key);
```
I know this code was like this already, but maybe we could change the loop to make it more obvious what it's doing?
```
    for (let i = node.index; i < node.index + colspan; i++) {
      let column = this.collection.columns[i];
      width += this.getColumnWidth(column.key);
```

snowystinger reviewed (commented):
General notes:
While resizing with the mouse, some rows that you aren't necessarily hovering will get what look like hover styles. The hover styles also flicker as resizing is going on which doesn't look great.

It'd be great to get more tests with the resize behaviors where the columns are resized in the test.

I think we've stayed non-breaking for the RSP component, not quite sure about the aria/stately users though

snowystinger commented on packages/@react-stately/table/src/useTableColumnResizeState.ts:
```diff
@@ -0,0 +1,181 @@
+    setColumnWidthsForRef(widths);
+  }, [columns, buildColumnWidths]);
+
+  function setColumnWidthsForRef(newWidths: Map<Key, number>) {
```
move this function up nearer to where the state and ref are defined so it's obvious how to set them both at the same time

snowystinger commented on packages/@react-stately/table/src/useTableColumnResizeState.ts:
```diff
@@ -0,0 +1,181 @@
+
+    // keep track of all columns that have been seized
+    resizedColumnsRef.current.add(column.key);
+    setResizedColumns(resizedColumnsRef.current);
```
this won't actually cause a re-render with state, resizedColumnsRef.current is still the same reference after `add`

do we need state for this? or is the ref enough? or do we need a new Set?

dnorth replied:
This is a really good question and a good observation. I believe at this time we don't want a re-render and the re-render happens later. If i'm not mistaken I think we followed this pattern of using both a ref and a `useState` because that's what was happening in a lot of other places, but you're right in saying that in this case it doesn't matter.

@marshallpete for the resizedColumns, especially because the `useState` isn't actually doing anything - should we just remove this `setResizedColumns` and only use the ref?

snowystinger commented on packages/@react-stately/table/src/useTableColumnResizeState.ts:
```diff
@@ -0,0 +1,181 @@
+  const [resizedColumns, setResizedColumns] = useState<Set<Key>>(new Set());
+  const resizedColumnsRef = useRef<Set<Key>>(resizedColumns);
+
+  let getRealColumnWidth = useCallback((column: GridNode<T>) => {
```
add type for returned value, i thought it was going to always be a number given the name

is this "Real" column width actually "default" or "user assigned" or "starting"?

Real to me sounds like the resolved width after all your maths

dnorth replied:
Haha this is a funny question because it's REALLY hard to name this. It really is the REAL one.

Real is the one that wins after a given order. Like it might be the width that was previously calculated, or if that doesn't exist then it might be user-assigned, it might be a default one assigned by the user, it might be a default one assigned by the developer, or it might be a default width assigned by the code. 

At the end of ALL of this, we have whatever the "real" width is. I'm not sure what else it could be named because it really is the ONE TRUE WIDTH depending on whatever that means at the time it's called, haha.

snowystinger replied:
sounds to me like it's `resolveColumnWidth`? or `getResolvedColumnWidth`?

and lets add a quick code comment about order of importantance when resolving that is easier to look at
`// resolve order: prevCalc -> userAssigned -> default user assigned -> default dev assigned -> default in code`

feel free to improve on it, but that looks easier to me than the ternary + nullish coalescing, I know that you can get that from the code, but it's nice to be able to read it in English i think, has a bit more context, i don't have to jump from "props" to "dev assigned"

snowystinger commented on packages/@react-stately/table/src/useTableColumnResizeState.ts:
```diff
@@ -0,0 +1,181 @@
+  let getRealColumnWidth = useCallback((column: GridNode<T>) => {
+    let props = column.props as ColumnProps<T>;
+    return resizedColumns?.has(column.key) ? columnWidthsRef.current.get(column.key) : props.width ?? props.defaultWidth ?? getDefaultWidth?.(column.props) ?? '1fr';
+  }, [getDefaultWidth, resizedColumns]);
```
```suggestion
  }, [getDefaultWidth, resizedColumns, props.width]);
```
is this missing props.width?

dnorth replied:
the dependency array es-lint rule gets mad at me if I add this because I think the function is still called but it doesn't need to regenerate the function because of a change in `props.width` or `props.defaultWidth`, right? If you think that the rule is wrong in this case then maybe we should talk about it so I can understand where this would break, so I can better understand useCallback

snowystinger replied:
o, i see what happened, i confused my props
there's the outer scope of props, and then there's the inner scope of props coming from column.props
lets rename `let props = column.props ...` to `let colProps` or something so I don't make that mistake again

snowystinger commented on packages/@react-stately/table/src/useTableColumnResizeState.ts:
```diff
@@ -0,0 +1,181 @@
+  function resizeColumn(column: GridNode<T>, newWidth: number) : { key: Key, width: number }[] {
+    // copy the columnWidths map and set the new width for the column being resized
+    let widths = new Map<Key, number>(columnWidthsRef.current);
+    widths.set(columnsRef.current[columnsRef.current.length - 1].key, 0);
+    widths.set(column.key, Math.max(
```
I don't understand this double widths.set, seems to be using the spooky column? but shouldn't that already be at 0?

dnorth replied:
@snowystinger yes you're right about the spooky column. @marshallpete could you add a comment here in the code explaining it? It might be helpful to pull out the `columnsRef.current[columnsRef.current.length - 1].key` into a variable so we know exactly what is being set to 0 and why.

marshallpete replied:
explanation in the code: [77c0b56](https://github.com/adobe/react-spectrum/pull/2883/commits/77c0b564176687c1574f44b6695e4167a3a451a7)

snowystinger commented on packages/@react-stately/table/src/useTableColumnResizeState.ts:
```diff
@@ -0,0 +1,181 @@
+      Math.min(Math.floor(newWidth), getMaxWidth(column.props.maxWidth, tableWidth.current))
+    ));
+
+    // keep track of all columns that have been seized
```
```suggestion
    // keep track of all columns that have been sized
```

snowystinger commented on packages/@react-stately/table/src/useTableColumnResizeState.ts:
```diff
@@ -0,0 +1,181 @@
+
+    // when getting recalculated columns above, the column being resized is not considered "recalculated"
+    // so we need to add it to the list of affected columns
+    let allAffectedColumns = new Map<Key, number>([[column.key, newWidth], ...recalculatedColumnWidths]);
+    return mapToArray(allAffectedColumns);
```
Can probably skip a couple steps here, in which case you can delete `mapToArray`

```suggestion
    return [[column.key, newWidth], ...recalculatedColumnWidths].map([key, width] => {key, width});
```

dnorth replied:
oh can you spread maps like this into just a regular array? I'll have to test that out to make sure that works correctly.

snowystinger replied:
Haha, I was assuming so because that's technically what you are already doing.
But you can check it out working here too, open the console and enter this:

```
const map1 = new Map();
map1.set('a', 1);
map1.set('b', 2);
map1.set('c', 3);
console.log([...map1])
```

dnorth replied:
I tested it and it did work! I made the change. I kept it named still just to make it clear what we're trying to get, but it definitely made it a bit more concise.

snowystinger reviewed (commented):
I think we'll need a story where the width of the Table can be changed as well

dnorth commented on packages/@react-stately/table/src/TableCollection.ts:
```diff
@@ -216,6 +217,30 @@ export class TableCollection<T> extends GridCollection<T> {
+      // this will be used to allow for smooth scrolling when
+      // resizing columns while a horizontal scroll bar is present
+      // (also lovingly known as the "spooky column")
+      let resizeBufferColumn: GridNode<T> = {
```
@marshallpete could you take care of this one? ☝️

marshallpete replied:
If the content of the table > table width, you get a horizontal scroll bar.
If you then scroll to the end of the content horizontally, and then try to shrink one of the columns by resizing it, the table content width gets smaller, changing how big the horizontal scroll bar is.
This causes some stuff to snap around and breaks the resize behavior.

To fix this, we add a "spooky column" to the end of the table. The width of this column defaults to 0. If you try and shrink a column and the width of the table contents > table width, then the "spooky column" will grow to take up the difference so that the total table content width remains constant while you are resizing. Once you finish resizing, the  "spooky column" snaps back to 0.

marshallpete replied:
https://user-images.githubusercontent.com/40001449/157966303-62256761-0f3b-4784-8a33-12b42b33aba7.mov

marshallpete replied:
Added comments to the code as well: [77c0b56](https://github.com/adobe/react-spectrum/pull/2883/commits/77c0b564176687c1574f44b6695e4167a3a451a7)

snowystinger commented on packages/@react-stately/table/src/useTableColumnResizeState.ts:
```diff
@@ -0,0 +1,181 @@
+  function onColumnResizeEnd(column: GridNode<T>, width: number) {
+    isResizing.current = false;
+    if (props.onColumnResizeEnd) {
+      let widthsObj = resizeColumn(column, width);
```
yes, i didn't specifically suggest adding a new ref for it because I was having trouble telling if another one was tracking something close to it or if you'd need a new one

dnorth replied:
Made this change, I think it is a lot better thanks for the suggestion.

snowystinger commented on packages/@react-stately/table/src/useTableColumnResizeState.ts:
```diff
@@ -5,6 +5,13 @@ import {GridNode} from '@react-types/grid';
 import {Key, MutableRefObject, useCallback, useRef, useState} from 'react';
 import {useLayoutEffect} from '@react-aria/utils';
 
+export interface AffectedColumnWidths {
```
interface is plural but entry is single, can we change it to either be `AffectedColumnWidth` or build the array into the type?

dnorth commented on packages/@react-stately/table/src/utils.ts:
```diff
@@ -0,0 +1,116 @@
+}
+    
+export function isStatic(width: number | string): boolean {
+  return width !== null && width !== undefined && (typeof width === 'number' || width.match(/^(\d+)%$/) !== null);
```
> I think I'm interpreting it as *fr/null/undefined are dynamic columns and numbers or percents are static?

Yes that is the case and I think our documentation is out of date because we used to say that percents were dynamic, when that is not true anymore.

dnorth replied:
btw I just realized why you did `width != undefined` instead of `width !== undefined` (I previously thought it was an accident) and so I went to make that change but it looks like the es-lint rule doesn't like that. So we should probably keep it explicitly `width !== null && width !== undefined` ? 

![image](https://user-images.githubusercontent.com/6720502/157776284-a1218242-756c-404a-a377-81b48360e04a.png)

snowystinger replied:
ah, sorry, `width != null` should pass lint then, I see we have that in a number of places in our code base, I never remember it off the top of my head, sorry for the confusion

dnorth commented on packages/@react-stately/table/src/utils.ts:
```diff
@@ -0,0 +1,116 @@
+
+export function parseStaticWidth(width: number | string, tableWidth: number): number {
+  if (typeof width === 'string') {
+    let match = width.match(/^(\d+)%$/);
```
This is actually an exact copy of what the code was before! But I think it would be good to update it.

dnorth commented on packages/@react-stately/table/src/utils.ts:
```diff
@@ -0,0 +1,116 @@
+      : 75;
+}
+
+function mapColumns<T>(remainingColumns: GridNode<T>[], remainingSpace: number, tableWidth: number): mappedColumn<T>[] {
```
Isn't the PDF in this PR already? I'm not sure how I can add it as an artifact to the PR - I know it's in a comment but maybe there's a more "official" way to add it that you're referring to?

snowystinger replied:
Hmmm I don't see it in the Files changed list. I think we can just put it in that way though. It'd probably live in the RFC directory https://github.com/adobe/react-spectrum/tree/main/rfcs

dnorth replied:
I see, so you mean you want it IN the codebase. I meant it was in a comment to the PR but I definitely never added it to the codebase.

I was actually thinking if we added it anywhere it would be in the docs but if you want it in the RFC directory I can put it there.

snowystinger commented on packages/@react-stately/table/src/utils.ts:
```diff
@@ -0,0 +1,113 @@
+        0
+      );
+    
+  let columns = [...remainingColumns].map((column, index) => {
```
```suggestion
  let columns = remainingColumns.map((column, index) => {
```
remainingColumns is already an array and `map` will return a new array, no need to make a copy

snowystinger commented on packages/@react-stately/table/src/utils.ts:
```diff
@@ -0,0 +1,113 @@
+export function getMinWidth(minWidth: number | string, tableWidth: number): number {
+  return minWidth != null
+      ? parseStaticWidth(minWidth, tableWidth)
+      : 75;
```
magic number? where is this coming from?

marshallpete replied:
75 is the pre-existing min. https://github.com/adobe/react-spectrum/blob/main/packages/%40react-stately/layout/src/TableLayout.ts#L97

snowystinger replied:
We should probably pull it out into a shared variable so we know this
it's either that or it needs to be sent in from the Layout since not every Layout will have this as the default min width

snowystinger replied:
Danny clarified, it was moved, it doesn't exist in two places. So I'm ok with this

snowystinger commented on packages/@react-stately/table/src/utils.ts:
```diff
@@ -0,0 +1,113 @@
+  return remainingColumns;
+}  
+    
+export function getDynamicColumnWidths<T>(remainingColumns: GridNode<T>[], remainingSpace: number, tableWidth: number) {
```
this isn't "remainingColumns" this is just all dynamic columns right?
and it's not "remainingSpace", it's "availableSpace" for the dynamic columns
is that correct? Or is this truly a subset of the dynamic columns and a subset of the available space for dynamic columns?

similar comments on some of the other functions in this file

snowystinger commented on packages/@react-stately/table/src/utils.ts:
```diff
@@ -0,0 +1,113 @@
+}
+
+function mapDynamicColumns<T>(remainingColumns: GridNode<T>[], remainingSpace: number, tableWidth: number): mappedColumn<T>[] {
+  let remainingFractions = remainingColumns.reduce(
```
this is totalFractions right?

marshallpete replied:
[7259f56](https://github.com/adobe/react-spectrum/pull/2883/commits/7259f5678facbc73a004c5defe9bd2cd9604b176)

snowystinger commented on packages/@react-stately/table/src/utils.ts:
```diff
@@ -0,0 +1,113 @@
+      index,
+      delta: Math.max(
+            0,
+            getMinWidth(column.props.minWidth, tableWidth) - targetWidth,
+            targetWidth - getMaxWidth(column.props.maxWidth, tableWidth)
```
just to make sure I'm thinking about this correctly
unless the target width is smaller than the min width
or unless the target width is greater than the max width

delta will always be 0 because the other two numbers will be negative?

marshallpete replied:
That's correct. Delta is how many pixels beyond the bounds (min/max) a column wants to be. It's the absolute value of the difference of clamped vs unclamped.

snowystinger commented on packages/@react-stately/table/src/utils.ts:
```diff
@@ -0,0 +1,113 @@
+  return columns;
+}
+
+function findDynamicColumnWidths<T>(remainingColumns: mappedColumn<T>[], remainingSpace: number, tableWidth: number): mappedColumn<T>[] {
```
```suggestion
function findDynamicColumnWidths<T>(mappedColumns: mappedColumn<T>[], remainingSpace: number, tableWidth: number): mappedColumn<T>[] {
```

snowystinger replied:
This function is a little confusing, it changes remainingColumns in place, but yet it returns remainingColumns, which makes it feel like it should be generating a new copy and not mutating the one passed in

I generally am not a fan of functions that side effect the arg passed in, so lets have your for loop become a map

dnorth commented on packages/@react-stately/table/package.json:
```diff
@@ -18,6 +18,7 @@
   },
   "dependencies": {
     "@babel/runtime": "^7.6.2",
+    "@react-aria/utils": "^3.11.2",
```
I forgot about this one. We should talk on monday about what we want to do here. I only used the one from `@react-aria/utils` because I got a warning to use it.

snowystinger replied:
Info on this one, we're leaving the circular in for the moment so we don't add an export to stately/utils that we don't really want in stately/utils. We'll be addressing this in followup. We think this particular case won't cause an issue or bugs because we've apparently accidentally done this before. We do want to remove it soon, but again, not necessary for getting this work into the hands of teams that need to try it out.

marshallpete commented on packages/@react-spectrum/table/src/Resizer.tsx:
```diff
@@ -0,0 +1,19 @@
+  let {resizerProps} = useTableColumnResize(state, item, ref);
+  return (
+    <FocusRing focusRingClass={classNames(styles, 'focus-ring')}>
+      <div ref={ref} {...resizerProps} className={classNames(styles, 'spectrum-Table-columnResizer')} role="separator" aria-orientation="vertical" aria-label="Resize column" />
```
aria-labeledby: [c2571b5](https://github.com/adobe/react-spectrum/pull/2883/commits/c2571b555b7970c6a0488036ec122be747c8e3cb)
aria-value*: [1999421](https://github.com/adobe/react-spectrum/pull/2883/commits/1999421f0fa2dd42e6df2cbdebd1aaab2567643c)

snowystinger commented on packages/@react-stately/table/src/utils.ts:
```diff
@@ -0,0 +1,113 @@
+}  
+    
+export function getDynamicColumnWidths<T>(remainingColumns: GridNode<T>[], remainingSpace: number, tableWidth: number) {
+  let columns = mapDynamicColumns(remainingColumns, remainingSpace, tableWidth);
+    
+  columns.sort((a, b) => b.delta - a.delta);
+  columns = findDynamicColumnWidths(columns, remainingSpace, tableWidth);
+  columns.sort((a, b) => a.index - b.index);
```
I think you were worried about sort times and whatnot, so I had a thought
I believe this is (2n [reduce + map] + nlogn (sort) + 2n [reduce + map] + nlogn (sort)) => 4n + 2nlogn
but you could also keep a priority queue (heaps work well) while doing mapDynamicColumns, as you come across columns where the delta is 0, you handle them, if it's > 0 you add it to the priority queue, once you've run through the list, you begin handling the queue
handle means do what you do in findDynamicColumnWidths

I believe this would become
(n [reduce] + nlog(n) [priority queue insertion] + n [anything in priority queue]) => 2n + nlogn
so in terms of extremely large sets, pretty much equivalent, but since this isn't extremely large, this might save a bit of time

this all hinges on me understanding the above functions well enough :) and you maybe be able to simplify the priority queue further, I don't quite know how that part factors in, but if it's just a bucket where we put everything that's going to be snapped to their bounds, then maybe this could be further reduced to O(n) 3n or linear

marshallpete replied:
Hopefully this explains why we need a sort and not just a queue. The order of the delta values is important. We could potentially put all the non-zero delta values into a separate array and only sort them and then concat all the 0 delta values onto the end. This would make it so we are only sorting clamped columns which should be a very small subset. In practice though, I don't think this would actually buy us much.
Question would be is the slightly more complex code worth the performance differece?

marshallpete commented on packages/@react-stately/table/src/useTableColumnResizeState.ts:
```diff
@@ -108,6 +112,8 @@ export default function useTableColumnResizeState<T>(props: ColumnResizeStatePro
   }
 
   function onColumnResizeEnd(column: GridNode<T>, width: number) {
+    console.log(tableWidth.current);
```
removed

marshallpete commented on packages/@react-spectrum/table/src/TableView.tsx:
```diff
@@ -408,27 +430,36 @@ function TableHeader({children, ...otherProps}) {
     isVirtualized: true
   }, state, ref);
 
+  let {buttonProps} = useButton(props, ref);
```
Added element type and made the button props conditional: [38ef25f](https://github.com/adobe/react-spectrum/pull/2883/commits/38ef25f1463cad02f4d64d507134c8e782114e2f)

NOTE: I added an interface that wasn't needed so I remove it in this commit: [21d6ea2](https://github.com/adobe/react-spectrum/pull/2883/commits/21d6ea291613a8f1f8562a5edc6569ef6569eda6)

snowystinger commented on packages/@react-stately/virtualizer/src/Virtualizer.ts:
```diff
@@ -661,7 +663,9 @@ export class Virtualizer<T extends object, V, W> {
 
         let item = this.getItem(visibleLayoutInfos.get(key).key);
-        if (view.content === item) {
+        // always need to rerender columns so that the resizer stuff gets rerendered correctly
```
better to give a concrete example I think than "stuff"
```suggestion
        // always need to rerender columns so that the resizer aria-values get updated
```

devongovett commented on packages/@react-stately/virtualizer/src/ReusableView.ts:
```diff
@@ -31,7 +31,7 @@ export class ReusableView<T extends object, V> {
 
   /** The content currently being displayed by this view, set by the collection view. */
-  content: T;
+  content: any;
```
Curious why this and below switched to `any`?

marshallpete replied:
In Virtualizer I'm checking content.props.allowsResizing. This was throwing type errors because props doesn't exist on type `T`. I spent some time trying to figure out what I need to do to get it to work correctly and be typed right but wasn't having any luck and needed to work on some other stuff so I set it to `any`. This is the commit with the change: [1999421](https://github.com/adobe/react-spectrum/pull/2883/commits/1999421f0fa2dd42e6df2cbdebd1aaab2567643c)

I am open to suggestions on how to do the typing correctly so we can avoid using `any`

dannify commented on packages/@react-stately/layout/src/TableLayout_DEPRECATED.ts:
```diff
@@ -0,0 +1,426 @@
+  getDefaultWidth: (props) => string | number
+}
+
+export class TableLayout_DEPRECATED<T> extends ListLayout<T> {
```
I would do this the over way around and name the "new" hooks with a different name so that it isn't a breaking change for existing hooks.
I understand that the swap over would be easier when no longer under feature flag, but I'd rather avoid renaming the existing hooks if we can.

dnorth replied:
I will do whatever you want but in my experience when doing "NEW" or "V2", the old gets forgotten about and is harder to untangle in the future. Are you _sure_ you would like me to make that change? (Again, i'll do whatever you want but I just want to make absolutely sure it's what you want)

dannify replied:
leaving for now

snowystinger commented on packages/@adobe/spectrum-css-temp/components/table/index.css:
```diff
@@ -86,6 +86,42 @@ svg.spectrum-Table-sortedIcon {
+  justify-content: end;
+  box-sizing: border-box;
+  position: absolute;
+  top: 0px;
+  right: 0px;
+  width: 10px;
+  height: 100%;
```
```suggestion
  inset-block-start: 0px;
  inset-inline-end: 0px;
  inline-size: 10px;
  block-size: 100%;
```

snowystinger commented on packages/@adobe/spectrum-css-temp/components/table/index.css:
```diff
@@ -86,6 +86,42 @@ svg.spectrum-Table-sortedIcon {
+  cursor: col-resize;
+  user-select: none;
+
+  &.rtl {
+    left: 0px;
+    right: auto;
+  }
+  
+  &::after {
```
```suggestion
  &::after {
```

snowystinger commented on packages/@adobe/spectrum-css-temp/components/table/index.css:
```diff
@@ -86,6 +86,42 @@ svg.spectrum-Table-sortedIcon {
+    position: absolute;
+    display: block;
+    box-sizing: border-box;
+    width: 1px;
+    height: 100%;
```
```suggestion
    inline-size: 1px;
    block-size: 100%;
```

snowystinger commented on packages/@adobe/spectrum-css-temp/components/table/index.css:
```diff
@@ -86,6 +86,42 @@ svg.spectrum-Table-sortedIcon {
+  &:active,
+  &:focus {
+    &::after {
+      width: 2px;
```
```suggestion
      inline-size: 2px;
```

snowystinger reviewed (commented):
You shouldn't need a `.rtl` class, that should be taken care of by logical properties

LFDanLu reviewed (commented):
Initial review, doing a deep dive of the state, layout, and virtualizer changes next

dnorth commented on packages/@react-spectrum/table/stories/Table.stories.tsx:
```diff
@@ -1044,9 +1045,249 @@ storiesOf('TableView', module)
+      </TableView>
+    )
+  )
+  .add(
+    'allowsResizing, uncontrolled, sortable columns',
+    () => <AsyncLoadingExample isResizable />,
+    {chromatic: {disable: true}}
+  )
```
After our virtualizer changes we noticed that the speed of sorting increased significantly. We're hoping those changes fixed the problem but this is something we also care about deeply and so it's the first thing we're planning on testing once we get it in our codebase.

LFDanLu replied:
Sounds good, I'll give it a quick once over but def not blocker for merge

dnorth commented on packages/@react-stately/table/src/index.ts:
```diff
@@ -10,6 +10,8 @@
  */
 
+export * from './useTableColumnResizeState';
+export * from './utils';
```
We do use `getContentWidth` in the `useTableColumnResizeState.test.ts` file like you said, so do you recommend we remove this export and then import in our test file like so?

```
  import {getContentWidth} from '../src/utils';
  import {renderHook} from '@testing-library/react-hooks';
  import {useTableColumnResizeState} from '../';
```

LFDanLu replied:
Right, no need to export anything that users/other packages won't use. We can always export them in the future if the need arises but for now we should exclude it IMO

snowystinger reviewed (approved):
Ran chromatic as well

LFDanLu reviewed (approved):
LGTM, feel free to address my previous comments in a followup. Things seem to performing much better now (especially sorting), thanks for all the great work!
